core/src/wsv.rs

```diff
@@ -393,30 +395,49 @@ impl WorldStateView {
         self.asset(id).map_err(Into::into)
     }
 
-    // TODO: There could be just this one method `blocks` instead of
-    // `blocks_from_height` and `blocks_after_height`. Also, this
-    // method would return references instead of cloning blockchain
-    // but comes with the risk of deadlock if consumer of the iterator
-    // stores references to blocks
-    /// Returns iterator over blockchain blocks
-    ///
-    /// **Locking behaviour**: Holding references to blocks stored in the blockchain can induce
-    /// deadlock. This limitation is imposed by the fact that blockchain is backed by [`dashmap::DashMap`]
-    #[inline]
-    pub fn blocks(&self) -> crate::block::ChainIterator {
-        self.blocks.iter()
+    /// Load all blocks in the block chain from disc and clone them for use.
+    #[allow(clippy::expect_used)]
+    pub fn all_blocks_by_value(&self) -> impl Iterator<Item = VersionedCommittedBlock> + '_ {
+        let block_count = self.block_hashes.borrow().len() as u64;
+        (1..=block_count)
+            .map(|height| {
+                self.kura
+                    .get_block_by_height(height)
+                    .expect("Failed to load block.")
+            })
+            .map(|block| VersionedCommittedBlock::clone(&block))
+    }
+
+    /// Load all blocks in the block chain from disc and clone them for use.
+    #[allow(clippy::expect_used)]
+    pub fn all_blocks_by_value_rev(&self) -> impl Iterator<Item = VersionedCommittedBlock> + '_ {
+        let block_count = self.block_hashes.borrow().len() as u64;
+        (1..=block_count)
+            .rev()
+            .map(|height| {
+                self.kura
+                    .get_block_by_height(height)
+                    .expect("Failed to load block.")
+            })
+            .map(|block| VersionedCommittedBlock::clone(&block))
```

Comment:
As i see we need places where we need to reverse order.

In this case i would suggest to return `impl DoubleEndedIterator<Item = VersionedCommittedBlock> + '_` from `all_blocks_by_value` and then anyone can just call `.rev()` and we can delete this method at all.

Reply:
Good suggestion.